mobile-sdk/plain-webview.md

```diff
@@ -0,0 +1,559 @@
+
+```js
+    window.webkit.messageHandlers.completed.postMessage("success")
+```
+```js
+    window.webkit.messageHandlers.canceled.postMessage()
+```
+
+#### Android
+
+{% include alert.html type="warning" icon="warning" header="Security Warning" body="Never use `WebView.addJavascriptInterface` on Android versions earlier than 4.2 (`Build.VERSION_CODES.JELLY_BEAN_MR1`)!" %}
+
+On Android, JavaScript interfaces are added by the `WebView.addJavascriptInterface` method. Any public methods with the `@JavascriptInterface` annotation of the passed-in object are exposed in JavaScript.
+
+{:.code-header}
+**Android**
+```kotlin
+    webView.addJavascriptInterface(
+        MyJsInterface(),
+        "callbacks"
+    )
+```
+```kotlin
+    class MyJsInterface {
+        // IMPORTANT!
+        // Methods annotated with @JavascriptInterface are
+        // NOT called on the main thread. They are called on
+        // a private, background, WebView thread.
+        // Make sure to only call methods that are safe
+        // to call in a background thread, or move execution to
+        // the main thread, e.g. by ViewModel.viewModelScope
+        // or LifecycleOwner.lifecycleScope (remember that
+        // FragmentActivity and Fragment implement LifecycleOwner).
+
+        @JavascriptInterface
+        fun completed(status: String) {
+            // Handle payment completion (success/failure)
+        }
+
+        @JavascriptInterface
+        fun canceled() {
+            // Handle payment cancellation
+        }
+    }
+```
+
+On Android, the objects added by `WebView.addJavascriptInterface` are exposed as globals with the specified name, and their `@JavascriptInterface public` methods with their JVM names (N.B! Be careful not to break the JVM names with Proguard or similar). Thus, your `completeUrl` and `cancelUrl` pages would need to eventually execute code like
+
+```js
+    callbacks.completed("success")
+```
+```js
+    callbacks.canceled()
+```
+
+## External Applications
+
+Before starting to implement lauching external applications, you should try to get at least one card payment working. With completion observing in place, you should be able to complete a payment flow, at least using the External Integration environment and its test cards.
+
+Sometimes, a payment flow calls for launching an external application, like BankID or Swish. A web page does this by opening a url that is handled by the app in question. To accommodate for this, we extend the "Listening for Navigations" approach above. If you opted for JavaScript hooks for completion, you will now need to add a navigation listener for external apps.
+
+Determining whether a url should launch an external app is straightforward, though on Android it involves a bit of a judgement call. Let us take a look at the arguably simpler iOS case first.
+
+### iOS
```

Comment:
Codacy found an issue: [[no-duplicate-headings] Do not use headings with similar content (186:1)](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=5758572)